src/ppx_deriving_qcheck/ppx_deriving_qcheck.ml

```diff
@@ -381,66 +351,123 @@ and gen_from_arrow ~loc ~env left right =
+    - `Normal of expression:
+    The derived generator is not recursive, we return only the generator.
+
+    - `Recursive of expression * expression
+    The derived generator was recursive (i.e. val gen : n -> t Gen.t), we return
+    the sized generator version, and a normal generator using this last with
+    [Gen.sized].
+*)
+let gen_from_type_declaration ~loc ~env td =
+  let (module A) = Ast_builder.make loc in
+  let ty = env.curr_type in
+  let is_rec = List.mem env.curr_type env.rec_types in
 
   let args =
     List.map
       (fun (typ, _) ->
-        match typ.ptyp_desc with Ptyp_var s -> pat ~loc s | _ -> assert false)
+        match typ.ptyp_desc with
+        | Ptyp_var s -> (pat ~loc s, name s |> A.evar)
+        | _ -> assert false)
       td.ptype_params
   in
+  let (args_pat, args_expr) = List.split args in
 
   let gen =
     match td.ptype_kind with
-    | Ptype_variant xs -> gen_from_kind_variant ~loc ~env name xs
+    | Ptype_variant xs ->
+       let is_rec cd = is_rec_constr_decl env.curr_types cd in
+       gen_sized ~loc is_rec (gen_from_constr ~loc ~env) xs
     | Ptype_record xs ->
         let gens = List.map (fun x -> gen_from_type ~loc ~env x.pld_type) xs in
-        record ~loc ~gens xs
+        gen_record ~loc ~gens xs
     | _ ->
         let typ = Option.get td.ptype_manifest in
-        gen_from_type ~loc ~env ~typ_name:name typ
+        gen_from_type ~loc ~env typ
   in
-  let gen = curry_args ~loc args gen in
 
-  [%stri let [%p pat_gen] = [%e gen]]
+  let pat_gen = pat ~loc ty in
+  if not is_rec then
+    let gen = Args.curry_args ~loc args_pat gen in
+    `Normal [%stri let [%p pat_gen] = [%e gen]]
+  else
+    let gen = Args.curry_args ~loc (args_pat @ [A.pvar "n"]) gen in
+    let pat_gen_sized = pat ~loc ~sized:true ty in
+    let gen_sized = name ~sized:true ty |> A.evar in
+    let gen_normal =
+      Args.curry_args ~loc args_pat
+        (G.sized ~loc (Args.apply_args ~loc args_expr gen_sized))
+    in
+    `Recursive (
+        [%stri let rec [%p pat_gen_sized] = [%e gen]],
+        [%stri let [%p pat_gen] = [%e gen_normal]]
+      )
 
-let derive_gen ~loc xs : structure =
+(** *)
+let mutually_recursive_gens ~loc gens =
+  let (module A) = Ast_builder.make loc in
+  let to_mutualize_gens =
+    List.map (function
+        | `Recursive (x, _) -> x
+        | `Normal x -> x) gens
+  in
+  let normal_gens =
+    List.filter_map (function
+        | `Recursive (_, x) -> Some x
+        | `Normal _ -> None) gens
+  in
+  let gens =
+    List.map (function
+        | [%stri let [%p? pat] = [%e? expr]]
+          | [%stri let rec [%p? pat] = [%e? expr]] ->
+           A.value_binding ~pat ~expr
+        | _ -> assert false) to_mutualize_gens
+  in
+  let mutual_gens = A.pstr_value Recursive gens in
+  mutual_gens :: normal_gens
+
+(** [derive_gen ~loc xs] creates generators for type declaration in [xs].
+
+    It also has a hidden purpose: it sets the environment prior to
```

Comment:
I think the hidden side-effect purpose is gone with the `env`-passing. 
If you agree, then I think the comment should be updated.

Reply:
Absolutely yes 0c71322b7fac81a33afe8387bbaabe08c2938831